docs/adr/0003-delivery-channels-templating.md

```diff
@@ -0,0 +1,92 @@
+# Templating engine for different notification types
+
+## Status
+
+Proposed
+
+## Context
+
+This ADR proposes a method on how to implement custom templating when a certain notification is sent through a communication channels (SMS, email, web push notifications etc).
+
+In a typical notification system, an object will be applied to a basic template to provide an end user with information targeting them. A lot of the basic design/layout of the template will be quite standard, but there will be places for values that will need to be inserted to customise it for the said end user. An example of this would be an birthday email notification, where a lot of the design and formatting of the email/subject will already be very well defined, and then insertion of custom data (eg. email id, name, etc ) is what will be needed by a templater.
+
+The content of this ADR will work off themes from the previous ADR (0002-delivery-channel-creation-and-integration.md).
+
+## Decision
+
+A notification having necessary information needs to be translated into a certain communication object (SMS, Email, etc) before it can be sent via the desired delivery channel (SendGrid, Twillio, Airship).
+
+An example of this would be a SMS, which needs the essential information to be available for it to be converted into an SMS object before being sent to the end user.
+
+```json
+
+    public class SMS {
+        public string ToPhoneNumber { get; set; }
+        public string Body { get; set; }
+    }
+
+```
+
+Each field in the communication object will needs it's own template (property) for the associated notification type. A template for the `ToPhoneNumber` property in the above example would look something like:
```

Comment:
What's the significance of "(property)" in brackets here? If I understand what you're saying, we'll have a template for each of the communication object's fields. I'm not sure what that "(property)" in the text is adding to this.

I'm also wondering whether a template really is the right mechanism to use for things like the target of a message (the phone number in this case, and presumably the `To:` address for email, etc.). I thought from the earlier ADR that the notification configuration system was meant to know the necessary details for how to send a particular kind of message to a user. So if a particular user has signed up to receive messages by SMS, they will have provided their phone number as part of that, and so the notification system has a record of which phone number to use when sending messages to that particular use by SMS. Requiring the target phone number to be filled in by a template seems like extra complication, and an additional opportunity for things to go wrong. What's the benefit of this?

Reply:
Agreed. I have updated the ADR.